File: JavaScript/LinkedList.js
/*
 * DESCRIPTION: Implementation of a linked list in JavaScript.  If you want to test it
 */

var LinkedList = function(){
    var list = {};
    list.size = 0;
    list.head = null;
    list.tail = null;

    //Adds item to the head
    list.addToHead = function(value){
        var newNode = createNode(value);
        newNode.next = list.head;
        list.head = newNode;
        if(list.tail == null){
            list.tail = newNode;
        }
        list.size++;
    };
    
    //Adds item to the tail
    list.addToTail = function(value){
        var newNode = createNode(value);
        if(tail == null){
            list.head = newNode;
            list.tail = newNode;
        }
        else{
            tail.next = newNode;
            tail = newNode;
        }
        list.size++;
    };

    //Removes item from the head
    list.removeHead = function(){
        if(list.isEmpty()){
            return;
        }
        list.head = list.head.next;
        list.size--;
        if(list.size == 0){
            list.tail = null;
        }
    };
    
    //Removes item from the tail
    list.removeTail = function(){
        if(list.isEmpty()){
            return;
        }
        var current = list.head;
        while(current.next.next != null){
            current = current.next;
        }
        current.next = current.next.next;
        list.size--;
    };

    //Checks if an item is in the list
    list.contains = function(key) {
        var current = list.head;
        while(current != null){
            if(current.data === key){
                return true;
            }
            else{
                current = current.next;
            }
        }
        return false;
    };

    //Prints every item in the list
    list.printList = function(){
        var current = list.head;
        while(current != null){
            console.log(current.data);
            current = current.next;
        }
    };

    //Checks if list is empty, if it is it prints a message
    list.isEmpty = function(){
      if(list.size == 0){
          console.log("List is empty!");
          return true;
      }
        return false;
    };

    return list;
};

var createNode = function(value){
    var Node = {};
    Node.data = value;
    Node.next = null;
    return Node;
};

//var list = new LinkedList();

File: CSharp/LinkedLists/LinkedList.cs
using System;
namespace MyLinkedList {
    class LinkedList<T> {
        public class Node{
            public T data;
            public Node next;
            public Node(T data){
                this.data = data;
            }
        }
        Node head {get; set;}
        Node tail {get; set;}
        int size {get; set;}

        /* Methods to add */
        public void AddHead(T data){
            Node newNode = new Node(data);
            newNode.next = head;
            head = newNode;
            if(tail == null){
                tail = newNode;
            }
            size++;
        }
        public void AddTail(T data){
            Node newNode = new Node(data);
            if(tail == null){
                head = newNode; 
                tail = newNode;
            }
            else{
                tail.next = newNode;
                tail = newNode;
            }
            size++;
        }

        /* Methods to delete */
        public T RemoveHead(){
            if(IsEmpty()){
                throw new Exception("List is empty!");
            }
            T data = head.data;
            head = head.next;
            if(head == null){
                tail = null;
            }
            size--;
            return data;
        }
        public T RemoveTail(){
            if(IsEmpty()){
                throw new Exception("List is empty!");
            }
            if(size == 1){
                return RemoveHead();
            }
            Node current = head;
            while(current.next.next != null){
                current = current.next;
            }
            T data = current.next.data;
            current.next = current.next.next;
            size--;
            return data;
        }

        /* Utility Methods */
        public bool IsEmpty(){
            return size == 0;
        }
        public void print(){
            if(IsEmpty()){
                return;
            }
            Node current = head;
            while(current != null){
                Console.WriteLine(current.data);
                current = current.next;
            }
        }
    }
}
